src/ptr.rs

```diff
@@ -65,92 +65,52 @@ pub struct Ptr<F: LurkField> {
 }
 
 #[allow(clippy::derived_hash_with_manual_eq)]
-impl<F: LurkField> Hash for Ptr<F> {
+impl<F: LurkField, T: Tag> Hash for GPtr<F, T> {
     fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
         self.tag.hash(state);
         self.raw.hash(state);
     }
 }
 
-impl<F: LurkField> Ptr<F> {
-    // TODO: Make these methods and the similar ones defined on expression consistent, probably including a shared trait.
-
-    // NOTE: Although this could be a type predicate now, when NIL becomes a symbol, it won't be possible.
-    /// check if a Ptr is `Nil` pointer
-    pub const fn is_nil(&self) -> bool {
-        matches!(self.tag, ExprTag::Nil)
-        // FIXME: check value also, probably
-    }
-
-    /// check if a Ptr is a `Cons` pointer
-    pub const fn is_cons(&self) -> bool {
-        matches!(self.tag, ExprTag::Cons)
-    }
-
-    // TODO: Is this still needed?
-    /// check if a Ptr is atomic pointer
-    pub const fn is_atom(&self) -> bool {
-        !self.is_cons()
-    }
-
-    // check if a Ptr is a list pointer
-    pub const fn is_list(&self) -> bool {
-        matches!(self.tag, ExprTag::Nil | ExprTag::Cons)
-    }
+pub type Ptr<F> = GPtr<F, ExprTag>;
+pub type ContPtr<F> = GPtr<F, ContTag>;
 
+impl<F: LurkField, T: Tag> GPtr<F, T> {
     /// check if a Ptr is an opaque pointer
     pub const fn is_opaque(&self) -> bool {
         self.raw.is_opaque()
     }
 
-    // TODO: Is this still needed?
-    pub const fn as_cons(self) -> Option<Self> {
-        if self.is_cons() {
-            Some(self)
-        } else {
-            None
-        }
-    }
-
-    // TODO: Is this still needed?
-    pub const fn as_list(self) -> Option<Self> {
-        if self.is_list() {
-            Some(self)
-        } else {
-            None
-        }
-    }
-
     /// Construct a Ptr from an index
-    pub fn index(tag: ExprTag, idx: usize) -> Self {
-        Ptr {
+    pub fn index(tag: T, idx: usize) -> Self {
+        GPtr {
             tag,
             raw: RawPtr::Index(idx),
             _f: Default::default(),
         }
     }
 
     /// Construct a Ptr from an opaque index
-    pub fn opaque(tag: ExprTag, idx: usize) -> Self {
-        Ptr {
+    pub fn opaque(tag: T, idx: usize) -> Self {
+        GPtr {
             tag,
             raw: RawPtr::Opaque(idx),
             _f: Default::default(),
         }
     }
 
     /// Construct a null Ptr
-    pub fn null(tag: ExprTag) -> Self {
-        Ptr {
+    pub fn null(tag: T) -> Self {
+        GPtr {
             tag,
             raw: RawPtr::Null,
             _f: Default::default(),
         }
     }
 
     #[inline]
-    pub fn cast(self, tag: ExprTag) -> Self {
-        Ptr {
+    pub fn cast(self, tag: T) -> Self {
```

Comment:
Does it really make sense to support casting of continuation pointers?

Reply:
Yup, that's a mistake. Thanks for the catch.